[content-visibility] Add support for css content-visibility: auto

rwlbuis opened the pull request:
#### b39fee5ba22e0f62398d7e869e271415f3d6ec84
<pre>
[content-visibility] Add support for css content-visibility: auto
<a href="https://bugs.webkit.org/show_bug.cgi?id=236711">https://bugs.webkit.org/show_bug.cgi?id=236711</a>

Reviewed by Tim Nguyen.

This patch implements support for content-visibility: auto [1].

To keep track of whether an element is relevant to the user [2] and
decide to whether it should skip contents [3] the following is done:
- state is kept using ContentRelevancyStatus enum class whether the
  content-visibility: auto subtree contains any visible, focused or
  selected elements, since that means it is relevant to the user.
- an IntersectionObserver is installed as soon as content-visibility: auto
  is set on an Element. This keeps track whether the Element is in the
  viewport or not and thus if it is relevant to the user.
- updating of c-v: auto elements relevancy and its effect on its subtree
  is done as part of &quot;Update the Rendering&quot;, see [4].

The methods focus() [4.3.6] and scrollIntoView() [4.3.5] treat any previously
skipped content as becoming relevant to the user before doing any scrolling.

Handling selected elements will be done in a future patch.

[1] <a href="https://drafts.csswg.org/css-contain/#valdef-content-visibility-auto">https://drafts.csswg.org/css-contain/#valdef-content-visibility-auto</a>
[2] <a href="https://drafts.csswg.org/css-contain/#relevant-to-the-user">https://drafts.csswg.org/css-contain/#relevant-to-the-user</a>
[3] <a href="https://drafts.csswg.org/css-contain/#skips-its-contents">https://drafts.csswg.org/css-contain/#skips-its-contents</a>
[4] <a href="https://drafts.csswg.org/css-contain/#cv-notes">https://drafts.csswg.org/css-contain/#cv-notes</a> (point 3)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-048-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-068-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-070-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-071-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-072-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-popover-top-layer-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-popover-top-layer-and-auto-descendant-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-popover-top-layer-and-auto-descendant-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-popover-top-layer-and-auto-descendant.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-and-auto-descendant-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-and-auto-descendant-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-and-auto-descendant.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-012-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp: Added.
(WebCore::ContentVisibilityDocumentState::observe):
(WebCore::ContentVisibilityDocumentState::unobserve):
(WebCore::ContentVisibilityDocumentState::intersectionObserver):
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements): integration into &quot;Update the Rendering&quot;, see [4]
(WebCore::ContentVisibilityDocumentState::updateContentRelevancyStatusForScrollIfNeeded): helper around lack of scroll anchoring.
(WebCore::ContentVisibilityDocumentState::updateOnScreenObservationTarget):
* Source/WebCore/dom/ContentVisibilityDocumentState.h: Added.
(WebCore::ContentVisibilityDocumentState::hasObservationTargets const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setFocusedElement): handle all focus changes.
(WebCore::Document::contentVisibilityDocumentState): keep track of observers for content-visibility: auto elements.
(WebCore::Document::isObservingContentVisibilityTargets const):
(WebCore::Document::updateRelevancyOfContentVisibilityElements): integration into &quot;Update the Rendering&quot;, see [4]
(WebCore::Document::scheduleContentRelevancyUpdate): integration into &quot;Update the Rendering&quot;, see [4]
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus): make sure scrolled to element is in visible subtree.
(WebCore::Element::hasFocusableStyle const): only test focus for c-v: hidden subtrees.
(WebCore::Element::scrollIntoView): make sure scrolled to element is in visible subtree.
(WebCore::Element::scrollIntoViewIfNeeded): make sure scrolled to element is in visible subtree.
(WebCore::Element::focus): make sure scrolled to element is in visible subtree.
(WebCore::Element::addToTopLayer): make sure element is in visible subtree.
(WebCore::Element::removeFromTopLayer): make sure element is not in visible subtree anymore.
(WebCore::Element::isRelevantToUser const): see [2].
(WebCore::Element::contentRelevancyStatus const):
(WebCore::Element::setContentRelevancyStatus):
(WebCore::Element::contentVisibilityViewportChange): handle changed visibility in viewport status.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::contentRelevancyStatus):
(WebCore::ElementRareData::setContentRelevancyStatus):
* Source/WebCore/dom/NodeRareData.h: keep track of whether content is relevant to the user.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::FrameView::maintainScrollPositionAtAnchor): handle anchor scrolls.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering): integration into &quot;Update the Rendering&quot;, see <a href="https://drafts.csswg.org/css-contain/#cv-notes">https://drafts.csswg.org/css-contain/#cv-notes</a>
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange): add/remove observation for content-visibility: auto element.
(WebCore::RenderElement::willBeDestroyed): remove observation for content-visibility: auto element.
(WebCore::RenderElement::isSkippedContentRoot const):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderObject::isSkippedContentRoot const):
* Source/WebCore/rendering/RenderElementInlines.h:
(WebCore::RenderElement::shouldApplyInlineSizeContainment const): take into account current value of effectiveSkippedContent.
(WebCore::RenderElement::shouldApplySizeContainment const): take into account current value of effectiveSkippedContent.
(WebCore::RenderElement::shouldApplySizeOrInlineSizeContainment const): take into account current value of effectiveSkippedContent.
(WebCore::RenderObject::isSkippedContentRoot const):
* Source/WebCore/rendering/RenderElementInlines.h:
(WebCore::RenderElement::shouldApplyInlineSizeContainment const):
(WebCore::RenderElement::shouldApplySizeOrInlineSizeContainment const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldSkipContent const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresLayout): a change in effectiveSkippedContent requires a layout.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::isSkippedContentRoot):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const): take content-visibility: auto into account.
* Source/WebCore/style/StyleSharingResolver.cpp:
(WebCore::Style::SharingResolver::canShareStyleWithElement const): do not share styles with different isRelevantToUser values.

Canonical link: <a href="https://commits.webkit.org/266084@main">https://commits.webkit.org/266084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98ef53ef1ac0c303f854d57f37c82ee4f0273d2f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12824 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13150 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13477 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14563 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12243 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12886 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15652 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13170 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14949 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12989 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13720 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10857 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15014 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11008 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/18674 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12083 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11776 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14974 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12227 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10139 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11490 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/11476 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3141 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15803 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12068 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -92,8 +92,10 @@ ExceptionOr<void> HTMLDialogElement::showModal()
+    if (!isInTopLayer()) {
+        updateContentRelevancyStatusIfNeeded(ContentRelevancyStatus::Visible, true);
         addToTopLayer();
+    }
```
Since the spec says to make _all_ top layer elements relevant, I think you want to move this `updateContentRelevancyStatusIfNeeded(ContentRelevancyStatus::Visible, true);` call to `addToTopLayer()`. Also do we need a reverse call in `removeFromTopLayer()`? content can remain visible, but irrelevant after removing from top layer (fullscreen for instance).

rwlbuis replied:
Good points, fixed.

nt1m replied:
Would be nice to write WPT for popovers/fullscreen. I think the current ones only cover dialog?

rwlbuis replied:
Great idea, I added the ones for popover:
https://github.com/web-platform-tests/wpt/pull/40541

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2247,4 +2261,13 @@ FloatRect RenderElement::referenceBoxRect(CSSBoxType boxType) const
+    if (style().contentVisibility() == ContentVisibility::Hidden)
+        return true;
+    return element() && !element()->isRelevantToUser();
+}
```
I think a switch(style().contentVisibility()) style would make it more obvious that isRelevantToUser belongs to ContentVisibility::Auto

rwlbuis replied:
Done.

alanbaradlay commented on Source/WebCore/rendering/RenderElementInlines.h:
```diff
@@ -73,6 +73,8 @@ inline bool RenderElement::shouldApplyAnyContainment() const
 {
+    if (shouldSkipContent())
+        return true;
     return shouldApplySizeOrStyleContainment(style().containsInlineSize());
```
maybe shouldSkipContent() || shouldApplySizeOrStyleContainment() ?

rwlbuis replied:
Done.

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -865,6 +873,9 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
         setHasReflection(false);
+    } else {
+        if (newStyle.contentVisibility() == ContentVisibility::Auto && element())
+            ContentVisibilityDocumentState::observe(*element());
```
It'd be great to find a common place for ContentVisibilityDocumentState::observer/unobserve logic here in styleWillChange

rwlbuis replied:
Good point, I did not like it either. I added a lambda for this.

alanbaradlay commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -541,6 +543,13 @@ inline RenderObject* RenderElement::lastInFlowChild() const
+{
+    if (isText())
+        return false;
+    return downcast<RenderElement>(*this).shouldSkipContent();
```
(nit) I am not too fond of this, but I assume it's a pattern we do.

alanbaradlay reviewed (commented):
rendering bits look ok. Someone from DOM side should look at the rest.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -0,0 +1,106 @@
+{
+    if (m_observer) {
+        for (auto target : m_observer->observationTargets()) {
+            if (target && target->renderer() && target->renderer()->style().contentVisibility() == ContentVisibility::Auto)
```
```suggestion
            if (target && target->renderStyle() && target->renderStyle().contentVisibility() == ContentVisibility::Auto)
```

rwlbuis replied:
I forgot about renderStyle(). Just testing renderer() is still quicker, so I kept that, but indeed it makes sense to replace renderer()->style() with renderStyle().

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5363,4 +5374,40 @@ bool Element::isPopoverShowing() const
+        ensureElementRareData().setContentRelevancyStatus(status, value);
+        invalidateStyle();
+    }
+}
```
It's a bit confusing that `updateContentRelevancyStatusIfNeeded` walks over the parent nodes when this method doesn't.

Also, can we set this up in a way that makes it clear when `updateContentRelevancyStatusIfNeeded` should be called, and when `updateContentRelevancyStatus` should be called?

E.g. if `updateContentRelevancyStatusIfNeeded` should always be called, it might be preferable to make this a static function.

rwlbuis replied:
I can see that the names can be confusing, and the "IfNeeded" part is more of an implementation detail, so I renamed these two methods to hopefully make things more clear.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5363,4 +5374,40 @@ bool Element::isPopoverShowing() const
+
+void Element::contentVisibilityViewportChange(bool isIntersecting)
+{
+    updateContentRelevancyStatus(ContentRelevancyStatus::Visible, isIntersecting);
```
Can we assert that the element has C-V auto? (Since this is called from the intersection observer, it must have C-V auto I'm guessing?)

question: Also, why aren't we iterating through the parent nodes like `updateContentRelevancyStatusIfNeeded` is doing?

rwlbuis replied:
You are right, this can only be called from c-v auto elements, since only those elements add/remove an observer. The assert makes sense, added.

We are not iterating here since there is no need, we know we are dealing with the c-v: auto element so we can just update the relevancy information directly.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5363,4 +5374,43 @@ bool Element::isPopoverShowing() const
+    for (auto target = this; target; target = target->parentElementInComposedTree()) {
+        if (target->renderer() && target->renderStyle()->contentVisibility() == ContentVisibility::Auto) {
+            target->setContentRelevancyStatus(status, value);
+            return;
```
What if one of the offscreen ancestors has `content-visibility: auto;` ? Is it accurate to early return here in that case?

rwlbuis replied:
For the non-nested situation, stopping at the first c-v: auto element is as intended.
For a nested c-v: auto situation, what I think you are describing, I am not sure the specification is strict there.
Anyway, I prefer to deal with this when adding support for ContentVisibilityAutoStateChangeEvent (https://bugs.webkit.org/show_bug.cgi?id=258880), where this is easier to test and existing should show up any problems or we can add new ones.
(FWIW right now removing the early return has no effect on the WPT test results).

nt1m replied:
Would be good to make testcases with nested `content-visibility: auto` and see if we match other browsers here.

rwlbuis replied:
I updated the PR and added content-visibility-auto-nested-scroll.html to test that. It turns out the early return has to be removed to match the other browsers for this test.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3892,6 +3901,8 @@ void Element::removeFromTopLayer()
     document().removeTopLayerElement(*this);
     clearNodeFlag(NodeFlag::IsInTopLayer);
 
+    updateContentRelevancyStatus(ContentRelevancyStatus::Visible, false);
```
This seems potentially hazardous to me. If the ancestor of the top layer element has C-V auto, but is on-screen, you actually do not want to set this to false on the ancestor element. Same if the element itself has C-V auto & remains visible after exiting the top layer.

Probably the most simple fix would be to add a separate `IsInTopLayer` reason for `ContentRelevancyStatus`.

(Also it would be nice to have a test for this too!)

rwlbuis replied:
Good point! I added the IsInTopLayer reason and a test (content-visibility-with-top-layer-in-auto-subtree-removal.html).

nt1m commented on Source/WebCore/rendering/RenderElementInlines.h:
```diff
@@ -98,11 +98,15 @@ inline bool RenderElement::shouldApplyPaintContainment() const
-    return shouldApplySizeOrStyleContainment(style().containsSize() || style().contentVisibility() == ContentVisibility::Hidden);
+    if (shouldSkipContent())
+        return true;
+    return shouldApplySizeOrStyleContainment(style().containsSize());
```
```suggestion
    return shouldSkipContent() || shouldApplySizeOrStyleContainment(style().containsSize());
```

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/rendering/RenderElementInlines.h:
```diff
@@ -98,11 +98,15 @@ inline bool RenderElement::shouldApplyPaintContainment() const
 {
+    if (shouldSkipContent())
+        return true;
     return shouldApplySizeOrStyleContainment(style().containsSizeOrInlineSize() || style().contentVisibility() == ContentVisibility::Hidden);
```
Since `shouldSkipContent()` already checks for C-V hidden, this works too:
```suggestion
    return shouldSkipContent() || shouldApplySizeOrStyleContainment(style().containsSizeOrInlineSize());
```

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/dom/Element.h:
```diff
@@ -105,6 +105,13 @@ struct ShadowRootInit;
 
+// https://drafts.csswg.org/css-contain/#relevant-to-the-user
+enum class ContentRelevancyStatus : uint8_t {
+    Visible = 1 << 0,
```
Can we rename this to `OnScreen` to match the spec ?

rwlbuis replied:
Sure, done.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5363,4 +5374,41 @@ bool Element::isPopoverShowing() const
+    for (auto target = this; target; target = target->parentElementInComposedTree()) {
+        if (target->renderer() && target->renderStyle()->contentVisibility() == ContentVisibility::Auto)
+            target->setContentRelevancyStatus(status, value);
+    }
```
Probably a followup, but there's probably room for optimizations here given we have the list of observation targets available. I'll leave it up to you how that optimization should exactly look like.

rwlbuis replied:
You are right, I added a FIXME.

nt1m commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -964,6 +965,21 @@ constexpr bool RenderStyle::preserveNewline(WhiteSpace mode)
     return mode != WhiteSpace::Normal && mode != WhiteSpace::NoWrap;
 }
 
+inline bool RenderStyle::determineEffectiveSkipsContent(const Element* element) const
```
Can we make this a static function? We shouldn't add this method to the `RenderStyle` class.

rwlbuis replied:
You meant a global function, right? Done.

nt1m commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -541,6 +543,13 @@ inline RenderObject* RenderElement::lastInFlowChild() const
     return nullptr;
 }
 
+inline bool RenderObject::shouldSkipContent() const
```
There are too many functions with similar names, which makes this confusing.

1. RenderObject::shouldSkipContent
2. RenderStyle::determineEffectiveSkipsContent
3. RenderObject::isSkippedContent
4. RenderStyle::effectiveSkipContents

Can we change these to something like:

1. RenderObject::isSkippedContentRoot
2. isSkippedContentRoot(Element* element, RenderStyle& style) (see other comment)
3. RenderObject::isSkippedContent
4. RenderStyle::isSkippedContent


I think 1 & 2 should share names, ditto for 3 & 4, since they represent the same thing.

rwlbuis replied:
I like shouldSkipContent as it is very clear, but it does state what should be done and not what it is. Given isInertSubtreeRoot exists, isSkippedContentRoot makes sense. Since it is a bit intrusive change I propose to do it in a separate PR (https://github.com/WebKit/WebKit/pull/15631).

rwlbuis replied:
Regarding, the other half, I am not so sure. effectiveSkipContents is a bit of a strange flag, it gels set on the root and gets inherited (likely same as effectiveInert). However then isSkippedContent needs to test the fag is true, but on the parent, to filter out the root. So I am not sure we can really map them. Let me know what you think, it is easy to add the renaming to https://github.com/WebKit/WebKit/pull/15631, which likely should land first.

nt1m replied:
I think `effectiveSkipContents` should be renamed to `effectiveSkippedContent` at least, because `SkipContents` implies it is the root skipped content (which is not true since the flag is inherited down)

rwlbuis replied:
> I think `effectiveSkipContents` should be renamed to `effectiveSkippedContent` at least, because `SkipContents` implies it is the root skipped content (which is not true since the flag is inherited down)

That seems reasonable, I made https://github.com/WebKit/WebKit/pull/15640 for that rename.

rwlbuis replied:
Renaming should now be done.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3892,6 +3901,8 @@ void Element::removeFromTopLayer()
     clearNodeFlag(NodeFlag::IsInTopLayer);
 
+    updateContentRelevancyStatus(ContentRelevancyStatus::IsInTopLayer, false);
+
```
I just realized there's another bug with top layer handling.

If you have a top layer descendant that has C-V auto, you'll likely want those to be updated as well, not just the ancestors.

rwlbuis replied:
Do you mean nested top layers? Can you give the pseudo html hierarchy?

nt1m replied:
I think your new patch fixes it, but would be worth a testcase:
```
<div style="height: 100vh"></div> (or whatever you need to make it offscreen)
<div popover="manual" style="display: block; position: static;" id="popover">
    <span style="content-visibility: auto">Hello</span>
 </div>
<script>
popover.showPopover();
</script>
```

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-resize-observer-no-error.html:
```diff
@@ -27,4 +29,6 @@
 
 <script>
 new ResizeObserver(() => {});
+requestAnimationFrame(takeScreenshot);
```
This doesn't seem right, can we get down to why we're failing here? (potentially by comparing to what other browsers do)

I suspect it's either the fallback when the initial C-V callback hasn't happened yet, but I might be wrong.

rwlbuis replied:
Yeah, I am experimenting a bit, there seem to be timing issues for WK1 test runner compared to WK2, some WK1 c-v tests seem to stop early compared to WK2 test run. When running mini-browser the end result of WK1 seems fine. So this part you can ignore for now (hopefully we can avoid making test changes just for WK1).

nt1m replied:
Oh if this is only for WK1, I would just skip the tests there, we don't need to support new features for WK1. 

There are known timing issues for rendering update steps in WK1

rwlbuis replied:
Oh, that is good to know! Would you happen to know a bug or bugs for this?

nt1m replied:
We have rdar://75508813 filed internally, no public bugzilla atm though

nt1m replied:
I filed https://bugs.webkit.org/show_bug.cgi?id=259075

rwlbuis replied:
I implemented https://drafts.csswg.org/css-contain/#cv-notes Step 4 so the WK1 test runner problem is gone now.
However next time I get WK1 problems I will remember the new bug, thanks!

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-012.html:
```diff
@@ -53,6 +53,7 @@
 (async function() {
   // Size normally.
   await nextRendering();
+  await nextRendering();
```
ditto here

rwlbuis replied:
Like above, you can ignore this part.

nt1m replied:
Can you undo this change?

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -0,0 +1,166 @@
+void ContentVisibilityDocumentState::unobserve(Element& element, Document& document)
+{
+    if (auto& observer = document.contentVisibilityDocumentState().m_observer)
+        observer->unobserve(element);
```
Can we clear the `contentRelevancyStatus` from the element here?

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.h:
```diff
@@ -0,0 +1,57 @@
+
+    RefPtr<IntersectionObserver> m_observer;
+
+    Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>> m_onScreenObservationTargets;
```
I would suggest using a `WeakHashSet` here, since it's just a Set that doesn't need to be ordered.

rwlbuis replied:
Done.

nt1m commented on Source/WebCore/page/Page.cpp:
```diff
@@ -1823,6 +1823,10 @@ void Page::updateRendering()
         }
     });
 
+    forEachDocument([] (Document& document) {
```
Can we make this a RenderingUpdateStep like the others?

rwlbuis replied:
Added.

nt1m reviewed (commented):
Looks mostly OK, thanks for updating the model! I have a few comments

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -0,0 +1,165 @@
+            if (relevancyToCheck.contains(ContentRelevancyStatus::Focused))
+                setRelevancyValue(ContentRelevancyStatus::Focused, target->hasFocusWithin());
+
+            auto hasTopLayerinSubtree = [](const Element& target) {
+                for (auto& element : target.document().topLayerElements()) {
+                    if (element->isDescendantOf(target))
+                        return true;
+                }
+                return false;
+            };
+            if (relevancyToCheck.contains(ContentRelevancyStatus::IsInTopLayer))
+                setRelevancyValue(ContentRelevancyStatus::IsInTopLayer, hasTopLayerinSubtree(*target));
```
Does this case work: https://github.com/WebKit/WebKit/pull/9096#discussion_r1258662305 ?

(Perhaps it does through the intersection observer, not sure)

rwlbuis replied:
I created the test here:
https://github.com/web-platform-tests/wpt/pull/40970

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -0,0 +1,165 @@
+        return found;
+    };
+
+    if (auto *scrollAnchorRoot = findSkippedContentRoot(scrollAnchor)) {
```
```suggestion
    if (auto* scrollAnchorRoot = findSkippedContentRoot(scrollAnchor)) {
```

rwlbuis replied:
Fixed.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -0,0 +1,165 @@
+
+    if (auto *scrollAnchorRoot = findSkippedContentRoot(scrollAnchor)) {
+        for (auto target : m_observer->observationTargets()) {
+            if (target && target->renderer() && target->renderStyle()->contentVisibility() == ContentVisibility::Auto)
```
Aren't all the observed renderers supposed to be content-visibility: auto ? Should this be an ASSERT?

rwlbuis replied:
Good point, I replaced part of the if with an ASSERT.

nt1m reviewed (approved):
Looks good assuming EWS is green.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -0,0 +1,171 @@
+{
+    if (auto& observer = document.contentVisibilityDocumentState().m_observer) {
+        observer->unobserve(element);
+        element.setContentRelevancyStatus({ });
```
Do we do want to do this unconditionally?

Also, I realize we may also want to remove from `m_onScreenObservationTargets` as well.

rwlbuis replied:
What would the condition be? This is called either on RenderElement removal or c-v: auto being removed, to me both times it makes sense to reset the relevancy.
Good point about m_onScreenObservationTargets, done.

nt1m replied:
Oh, I meant not conditioning resetting the element flag on `if (auto& observer = document.contentVisibilityDocumentState().m_observer) {`

rwlbuis replied:
Ah, I see :)Fixed.

nt1m commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -0,0 +1,171 @@
+
+void ContentVisibilityDocumentState::observe(Element& element)
+{
+    auto& state = element.document().contentVisibilityDocumentState();
```
Isn't this just referencing itself? Should we assert the element.document() here?

rwlbuis replied:
This is done this way since observe/unobserve are static methods.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-and-auto-descendant.html:
See my comments on WPT PR

rwlbuis replied:
I now imported the WPTs.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9736,6 +9741,37 @@ std::optional<uint64_t> Document::noiseInjectionHashSalt() const
+    if (m_contentVisibilityDocumentState->hasSameFrameObservationTargets()) {
+        updateIntersectionObservations();
+        m_contentVisibilityDocumentState->clearSameFrameObservationTargets();
+    }
```
Not super convinced this is the best way to handle this, I'd be curious to know what other browsers do

rwlbuis replied:
Yes, this part is tricky since it is not specified well. The situation is described well here:
https://github.com/w3c/csswg-drafts/issues/8542

Basically IIUC chromium makes this work by using sync intersection observer internally, which is not specced AFAIK and not something gecko or WebKit provide. Therefore chromium is the only supporting this, it is explained in https://github.com/w3c/csswg-drafts/issues/8542#issuecomment-1459175417

I propose to leave out this support in this PR and open a bug for it, fixing it once the issue is a bit more discussed/accepted. Note that the only test it has an effect on for the test result is auto-012.html, it conveniently would "fix" the WK1 issues but the real fix for that probably still is https://bugs.webkit.org/show_bug.cgi?id=259075.

nt1m replied:
> I propose to leave out this support in this PR and open a bug for it, fixing it once the issue is a bit more discussed/accepted. Note that the only test it has an effect on for the test result is auto-012.html,

This test seems to pass in Gecko, so it would be nice seeing why that is the case.

But I'm OK implementing the Chromium proposal separately though.

nt1m reviewed (approved):
Looks OK.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-012-expected.txt:
```diff
@@ -1,9 +1,9 @@
 
-PASS Sizing normally
+FAIL Sizing normally assert_equals: clientWidth expected 50 but got 100
```
@mrobinson Do you know how this subtest passes in Gecko without the sync intersection observer ? (or is there maybe one that I missed?)

mrobinson replied:
I have forgotten a lot of things, but my intuition is that this is passing due to `nextRendering()` awaiting on two calls to `requestAnimationFrame()`. I suspect that is why this can pass in Gecko.

rwlbuis committed:
[content-visibility] Add support for css content-visibility: auto
https://bugs.webkit.org/show_bug.cgi?id=236711

Reviewed by Tim Nguyen.

This patch implements support for content-visibility: auto [1].

To keep track of whether an element is relevant to the user [2] and
decide to whether it should skip contents [3] the following is done:
- state is kept using ContentRelevancyStatus enum class whether the
  content-visibility: auto subtree contains any visible, focused or
  selected elements, since that means it is relevant to the user.
- an IntersectionObserver is installed as soon as content-visibility: auto
  is set on an Element. This keeps track whether the Element is in the
  viewport or not and thus if it is relevant to the user.
- updating of c-v: auto elements relevancy and its effect on its subtree
  is done as part of "Update the Rendering", see [4].

The methods focus() [4.3.6] and scrollIntoView() [4.3.5] treat any previously
skipped content as becoming relevant to the user before doing any scrolling.

Handling selected elements will be done in a future patch.

[1] https://drafts.csswg.org/css-contain/#valdef-content-visibility-auto
[2] https://drafts.csswg.org/css-contain/#relevant-to-the-user
[3] https://drafts.csswg.org/css-contain/#skips-its-contents
[4] https://drafts.csswg.org/css-contain/#cv-notes (point 3)

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/resources/resource-files.json:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-048-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-068-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-070-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-071-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-072-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-popover-top-layer-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-popover-top-layer-and-auto-descendant-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-popover-top-layer-and-auto-descendant-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-popover-top-layer-and-auto-descendant.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-and-auto-descendant-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-and-auto-descendant-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-with-top-layer-and-auto-descendant.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/w3c-import.log:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-012-expected.txt:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp: Added.
(WebCore::ContentVisibilityDocumentState::observe):
(WebCore::ContentVisibilityDocumentState::unobserve):
(WebCore::ContentVisibilityDocumentState::intersectionObserver):
(WebCore::ContentVisibilityDocumentState::updateRelevancyOfContentVisibilityElements): integration into "Update the Rendering", see [4]
(WebCore::ContentVisibilityDocumentState::updateContentRelevancyStatusForScrollIfNeeded): helper around lack of scroll anchoring.
(WebCore::ContentVisibilityDocumentState::updateOnScreenObservationTarget):
* Source/WebCore/dom/ContentVisibilityDocumentState.h: Added.
(WebCore::ContentVisibilityDocumentState::hasObservationTargets const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setFocusedElement): handle all focus changes.
(WebCore::Document::contentVisibilityDocumentState): keep track of observers for content-visibility: auto elements.
(WebCore::Document::isObservingContentVisibilityTargets const):
(WebCore::Document::updateRelevancyOfContentVisibilityElements): integration into "Update the Rendering", see [4]
(WebCore::Document::scheduleContentRelevancyUpdate): integration into "Update the Rendering", see [4]
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::focus): make sure scrolled to element is in visible subtree.
(WebCore::Element::hasFocusableStyle const): only test focus for c-v: hidden subtrees.
(WebCore::Element::scrollIntoView): make sure scrolled to element is in visible subtree.
(WebCore::Element::scrollIntoViewIfNeeded): make sure scrolled to element is in visible subtree.
(WebCore::Element::focus): make sure scrolled to element is in visible subtree.
(WebCore::Element::addToTopLayer): make sure element is in visible subtree.
(WebCore::Element::removeFromTopLayer): make sure element is not in visible subtree anymore.
(WebCore::Element::isRelevantToUser const): see [2].
(WebCore::Element::contentRelevancyStatus const):
(WebCore::Element::setContentRelevancyStatus):
(WebCore::Element::contentVisibilityViewportChange): handle changed visibility in viewport status.
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ElementRareData.cpp:
* Source/WebCore/dom/ElementRareData.h:
(WebCore::ElementRareData::contentRelevancyStatus):
(WebCore::ElementRareData::setContentRelevancyStatus):
* Source/WebCore/dom/NodeRareData.h: keep track of whether content is relevant to the user.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::FrameView::maintainScrollPositionAtAnchor): handle anchor scrolls.
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering): integration into "Update the Rendering", see https://drafts.csswg.org/css-contain/#cv-notes
(WebCore::operator<<):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::styleWillChange): add/remove observation for content-visibility: auto element.
(WebCore::RenderElement::willBeDestroyed): remove observation for content-visibility: auto element.
(WebCore::RenderElement::isSkippedContentRoot const):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderObject::isSkippedContentRoot const):
* Source/WebCore/rendering/RenderElementInlines.h:
(WebCore::RenderElement::shouldApplyInlineSizeContainment const): take into account current value of effectiveSkippedContent.
(WebCore::RenderElement::shouldApplySizeContainment const): take into account current value of effectiveSkippedContent.
(WebCore::RenderElement::shouldApplySizeOrInlineSizeContainment const): take into account current value of effectiveSkippedContent.
(WebCore::RenderObject::isSkippedContentRoot const):
* Source/WebCore/rendering/RenderElementInlines.h:
(WebCore::RenderElement::shouldApplyInlineSizeContainment const):
(WebCore::RenderElement::shouldApplySizeOrInlineSizeContainment const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldSkipContent const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareInheritedDataChangeRequiresLayout): a change in effectiveSkippedContent requires a layout.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::isSkippedContentRoot):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const): take content-visibility: auto into account.
* Source/WebCore/style/StyleSharingResolver.cpp:
(WebCore::Style::SharingResolver::canShareStyleWithElement const): do not share styles with different isRelevantToUser values.

Canonical link: https://commits.webkit.org/266084@main

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
```diff
@@ -0,0 +1,176 @@
+// Workaround for lack of support for scroll anchoring. We make sure any content-visibility: auto elements
+// above the one to be scrolled to are already hidden, so the scroll position will not need to be adjusted
+// later.
+// FIXME: remove when scroll anchoring is implemented.
```
This needs a bug filed that's related to the scroll anchoring bug so we don't forget to remove this workaround.

rwlbuis replied:
Good point, I opened https://bugs.webkit.org/show_bug.cgi?id=259269 for this.

smfr commented on Source/WebCore/dom/ContentVisibilityDocumentState.h:
```diff
@@ -0,0 +1,58 @@
+
+    bool updateRelevancyOfContentVisibilityElements(const OptionSet<ContentRelevancyStatus>&);
+
+    void updateOnScreenObservationTarget(const Element&, bool);
```
I can't tell what the bool means.

rwlbuis replied:
I added an enum instead (change in on https://github.com/WebKit/WebKit/pull/15879).

smfr commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -9549,6 +9554,34 @@ std::optional<uint64_t> Document::noiseInjectionHashSalt() const
+{
+    if (!isObservingContentVisibilityTargets())
+        return;
+    if (m_contentVisibilityDocumentState->updateRelevancyOfContentVisibilityElements(m_contentRelevancyStatusUpdate))
```
How do you know m_contentVisibilityDocumentState is not null here?

nt1m replied:
`isObservingContentVisibilityTargets` checks for it

smfr commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -768,8 +769,13 @@ Vector<String> Element::getAttributeNames() const
 {
-    if (renderer() && renderer()->isSkippedContent())
-        return false;
+    if (renderer() && renderer()->isSkippedContent()) {
```
This block should be wrapped up in a function so it's easier to tell what it's doing.

rwlbuis replied:
I did not put much effort in this code since I think we should go with https://github.com/WebKit/WebKit/pull/15682, which simplifies Element::hasFocusableStyle.

smfr commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1066,6 +1072,8 @@ static std::optional<std::pair<RenderElement*, LayoutRect>> listBoxElementScroll
 
 void Element::scrollIntoView(std::optional<std::variant<bool, ScrollIntoViewOptions>>&& arg)
 {
+    document().contentVisibilityDocumentState().updateContentRelevancyStatusForScrollIfNeeded(*this);
```
This is making Document create a ContentVisibilityDocumentState even when it has no skipped content?

rwlbuis replied:
Good point, fixed.

smfr commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1131,6 +1139,8 @@ void Element::scrollIntoView(bool alignToTop)
 
 void Element::scrollIntoViewIfNeeded(bool centerIfNeeded)
 {
+    document().contentVisibilityDocumentState().updateContentRelevancyStatusForScrollIfNeeded(*this);
```
Ditto

rwlbuis replied:
Fixed.

smfr commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3520,6 +3530,8 @@ void Element::focus(const FocusOptions& options)
         return;
     }
 
+    document->contentVisibilityDocumentState().updateContentRelevancyStatusForScrollIfNeeded(*this);
```
Ditto

rwlbuis replied:
Fixed.

smfr commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -5363,4 +5379,28 @@ bool Element::isPopoverShowing() const
+// https://drafts.csswg.org/css-contain/#relevant-to-the-user
+bool Element::isRelevantToUser() const
+{
+    return hasRareData() && elementRareData()->contentRelevancyStatus();
```
Would be more readable with an explicit `!contentRelevancyStatus().isEmpty()`

smfr replied:
Also why doesn't this just call `contentRelevancyStatus()`?

rwlbuis replied:
Right, that is much simpler, done.

smfr commented on Source/WebCore/dom/Element.h:
```diff
@@ -105,6 +105,14 @@ struct ShadowRootInit;
 using ExplicitlySetAttrElementsMap = HashMap<QualifiedName, Vector<WeakPtr<Element, WeakPtrImplWithEventTargetData>>>;
 
+// https://drafts.csswg.org/css-contain/#relevant-to-the-user
+enum class ContentRelevancyStatus : uint8_t {
```
Not sure the word `Status` adds anything to the name.

rwlbuis replied:
Yeah it is a bit verbose, removed.

smfr commented on Source/WebCore/dom/ElementRareData.cpp:
```diff
@@ -36,6 +36,7 @@ namespace WebCore {
 struct SameSizeAsElementRareData : NodeRareData {
     unsigned short m_childIndex;
     int m_tabIndex;
+    uint32_t contentRelevancy;
```
But ContentRelevancyStatus  is a `uint8_t` ?

rwlbuis replied:
Oops, adjusted.

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -2407,6 +2408,9 @@ void LocalFrameView::maintainScrollPositionAtAnchor(ContainerNode* anchorNode)
     cancelScheduledScrolls();
 
+    if (is<Element>(anchorNode))
+        m_frame->document()->contentVisibilityDocumentState().updateContentRelevancyStatusForScrollIfNeeded(downcast<Element>(*anchorNode));
```
Forcing creation of ContentVisibilityDocumentState here again.

rwlbuis replied:
Fixed.

smfr commented on Source/WebCore/page/Page.h:
```diff
@@ -243,6 +243,7 @@ enum class RenderingUpdateStep : uint32_t {
     AccessibilityRegionUpdate       = 1 << 20,
 #endif
     RestoreScrollPositionAndViewState = 1 << 21,
+    UpdateContentRelevancy          = 1 << 22,
```
Please move this above IntersectionObservations.

rwlbuis replied:
Sure, moved.

smfr reviewed (commented):
In general I'm concerned about the multiple different terms being used in relation to content visibility. These are:

Content visibility
Skipped content
Content relevancy
Is relevant to user

Can we shrink the number of terms so it's easier to grep for all code related to this feature?
